pkg/manifests/manifests.go

```diff
@@ -761,7 +773,49 @@ func (f *Factory) KubeStateMetricsMinimalServiceMonitor() (*monv1.ServiceMonitor
 	return f.NewServiceMonitor(f.assets.MustNewAssetReader(KubeStateMetricsMinimalServiceMonitor))
 }
 
-func (f *Factory) KubeStateMetricsDeployment() (*appsv1.Deployment, error) {
+// Check if custom-resource-state metrics have been enabled.
+// * If so, add the --custom-resource-state-config-file flag to the kube-state-metrics container, if it is absent.
+// * If not, remove the --custom-resource-state-config-file flag from the kube-state-metrics container, if it is present.
+// This will, in turn, also cause the kube-state-metrics container to be restarted with the new set of flags in effect.
+func enableOrDisableCRSMetrics(d *appsv1.Deployment, enableCRSMetrics bool) {
```

Comment:
A separate function makes it harder to follow for me. Why not injecting the argument here?

https://github.com/openshift/cluster-monitoring-operator/blob/84be118af132567e6f84dfd6e72f3172877c5a3e/pkg/manifests/manifests.go#L774-L778

```go
		case "kube-state-metrics":
			d.Spec.Template.Spec.Containers[i].Image = f.config.Images.KubeStateMetrics
			if f.config.ClusterMonitoringConfiguration.KubeStateMetricsConfig.Resources != nil {
				d.Spec.Template.Spec.Containers[i].Resources = *f.config.ClusterMonitoringConfiguration.KubeStateMetricsConfig.Resources
			}
			if enableCRSMetrics {
				d.Spec.Template.Spec.Containers[containerIndex].Args = append(d.Spec.Template.Spec.Containers[containerIndex].Args, "--custom-resource-state-config-file=/etc/kube-state-metrics/custom-resource-state-configmap.yaml")
			}
		}
```

Reply:
Moving this out into its own function enabled me to test both of the `enableCRSMetrics` [branches](https://github.com/openshift/cluster-monitoring-operator/pull/2078/files#diff-1f9d2d18d4cabd8262e585b932330183f7cb9b544ac9f27c8bcb3ef9c8d14376R782-R816): https://github.com/openshift/cluster-monitoring-operator/pull/2078/files#diff-b388a528a3665a46f4dfa8eb4a8c02de116acb8f85e4611f12f11dc337d7f7b5R3564-R3568.

Reply:
Hmm, the static manifest doesn't have the flag in the first place and the factory loads that manifest so it's safe to assume that the argument will never present. To ensure that the assumption doesn't change in the future, we need an e2e test which is required anyway from my POV: there are many things that could go wrong and only e2e testing will validate that the metrics are ingested in Prometheus eventually.

FWIW this wouldn't be different from what we're doing elsewhere https://github.com/openshift/cluster-monitoring-operator/blob/e39355c5d44382b31937bf577da66883e835b01b/pkg/manifests/manifests.go#L2880-L2892

Reply:
Reverted.